core/src/main/java/com/sequenceiq/cloudbreak/service/stack/StackStopRestrictionService.java

```diff
@@ -9,68 +13,151 @@
-                if (onlyEphemeralStorageInInstanceGroup(instanceGroup)) {
-                    reason = StopRestrictionReason.EPHEMERAL_VOLUMES;
-                    LOGGER.info("Infrastructure cannot be stopped. Instances in group [{}] have ephemeral storage only.", instanceGroup.getGroupName());
-                    return reason;
-                } else if (isCbVersionBeforeStopSupport(stack)) {
-                    TemporaryStorage temporaryStorage = instanceGroup.getTemplate().getTemporaryStorage();
-                    if (TemporaryStorage.EPHEMERAL_VOLUMES.equals(temporaryStorage)) {
-                        reason = StopRestrictionReason.EPHEMERAL_VOLUME_CACHING;
-                        LOGGER.info("Infrastructure cannot be stopped. Group [{}] has ephemeral volume caching enabled. " +
-                                "Stopping clusters with ephemeral volume caching enabled are only available in clusters " +
-                                "created at least with Cloudbreak version [{}]", instanceGroup.getGroupName(), MIN_VERSION);
-                        return reason;
-                    }
-                    if (hasInstanceGroupAwsEphemeralStorage(instanceGroup)) {
-                        reason = StopRestrictionReason.EPHEMERAL_VOLUMES;
-                        LOGGER.info("Infrastructure cannot be stopped. Instances in group [{}] have ephemeral storage." +
-                                "Stopping clusters with ephemeral storage instances are only available in clusters " +
-                                "created at least with Cloudbreak version [{}]", instanceGroup.getGroupName(), MIN_VERSION);
-                        return reason;
-                    }
+        if (!config.getRestrictedCloudPlatform().equals(stack.getCloudPlatform())) {
+            return StopRestrictionReason.NONE;
+        }
+        for (InstanceGroup instanceGroup : stack.getInstanceGroups()) {
+            if (ephemeralVolumeChecker.instanceGroupContainsOnlyDatabaseAndEphemeralVolumes(instanceGroup)) {
+                StopRestrictionReason ephemeralResult = checkEphemeralOnlyInstanceGroupStoppable(instanceGroup, stack);
+                if (ephemeralResult != StopRestrictionReason.NONE) {
+                    return ephemeralResult;
+                }
+            } else {
+                StopRestrictionReason nonEphemeralResult = checkNonEphemeralOnlyInstanceGroupStoppable(instanceGroup);
+                if (nonEphemeralResult != StopRestrictionReason.NONE) {
+                    return nonEphemeralResult;
                 }
             }
         }
-        return reason;
+        return StopRestrictionReason.NONE;
+    }
+
+    private StopRestrictionReason checkEphemeralOnlyInstanceGroupStoppable(InstanceGroup instanceGroup, Stack stack) {
+        if (!isCbVersionBeforeStopSupport(stack, config.getEphemeralOnlyMinVersion())
+                || !isSaltComponentCbVersionBeforeStopSupport(stack.getCluster().getId())) {
```

Comment:
as talked about these shouldn't be invoked inside a for loop as it generates db calls for the same thing

Reply:
For the inject question: I've seen in some places that @ConfigurationProperties annotated components are used like this instead of using inject. Changed to use inject in the updated commit.